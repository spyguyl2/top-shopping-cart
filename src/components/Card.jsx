import { useState } from "react";
import PropTypes from "prop-types";
import SpinButton from "./SpinButton";
import "../styles/Card.css";

const Card = ({
  title,
  spinButtonId,
  image,
  price,
  onClick,
  min = 0,
  max = 99,
}) => {
  const [quantity, setQuantity] = useState(0);

  const handleIncreaseQuantity = () => {
    quantity >= max
      ? setQuantity(max)
      : setQuantity((prevQuantity) => parseInt(prevQuantity + 1));
  };
  const handleDecreaseQuantity = () => {
    quantity <= min
      ? setQuantity(min)
      : setQuantity((prevQuantity) => parseInt(prevQuantity - 1));
  };

  const handleInputChange = (event) => {
    let newValue = event.target.value;
    if (newValue <= min) newValue = min;
    if (newValue >= max) newValue = max;
    if (newValue === "") newValue = 0;
    setQuantity(newValue);
  };

  return (
    <div className="card">
      <h4>{title}</h4>
      <img src={image} alt="" />
      <p>{`$${price}`}</p>
      <SpinButton
        id={spinButtonId}
        min={min}
        max={max}
        handleIncreaseQuantity={handleIncreaseQuantity}
        handleDecreaseQuantity={handleDecreaseQuantity}
        handleInputChange={handleInputChange}
        quantity={quantity}
      />
      <button onClick={onClick}>Add To Cart</button>
    </div>
  );
};

Card.propTypes = {
  title: PropTypes.string.isRequired,
  spinButtonId: PropTypes.number.isRequired,
  image: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  onClick: PropTypes.func,
  min: PropTypes.number,
  max: PropTypes.number,
};

export default Card;
