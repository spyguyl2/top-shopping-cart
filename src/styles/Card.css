.card {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 0.8rem;
  width: 10rem;
  border: 1px solid purple;
  border-radius: 5px;
  background-color: rgb(180, 120, 237);
}
