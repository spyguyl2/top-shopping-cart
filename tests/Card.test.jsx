import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import Card from "../src/components/Card";
import userEvent from "@testing-library/user-event";

describe("Card component", () => {
  it("Renders component", () => {
    const container = render(<Card />);
    expect(container).toMatchSnapshot();
  });

  it("Accepts typed numbers", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const input = screen.getByRole("spinbutton");

    await user.click(input);
    await user.keyboard("45");

    expect(input.value).toMatch("45");
  });

  it("Does not accept other characters", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const input = screen.getByRole("spinbutton");

    await user.click(input);
    await user.keyboard("dsf@#");

    expect(input.value).toMatch("0");
  });

  it("Returns to min when a number below min is entered", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const input = screen.getByRole("spinbutton");

    await user.keyboard("45");
    await user.keyboard("-50");

    expect(input.value).toMatch("0");
  });

  it("Returns to max when a number above max is entered", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const input = screen.getByRole("spinbutton");

    await user.click(input);
    await user.keyboard("45");
    await user.keyboard("101");

    expect(input.value).toMatch("99");
  });

  it("Increases input value by 1 when plus is clicked", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const plusButton = screen.getByRole("button", { name: "plus" });

    await user.click(plusButton);

    expect(screen.getByRole("spinbutton").value).toMatch("1");
  });

  it("Decreases input value by 1 when minus is clicked", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const minusButton = screen.getByRole("button", { name: "minus" });
    const input = screen.getByRole("spinbutton");

    await user.click(input);
    await user.keyboard("2");
    await user.click(minusButton);

    expect(input.value).toMatch("1");
  });

  it("Does not go under the min value when minus is clicked.", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const minusButton = screen.getByRole("button", { name: "minus" });
    const input = screen.getByRole("spinbutton");

    await user.click(minusButton);

    expect(input.value).toMatch("0");
  });

  it("Does not go over the max value when plus is clicked", async () => {
    const user = userEvent.setup();

    render(<Card />);
    const plusButton = screen.getByRole("button", { name: "plus" });
    const input = screen.getByRole("spinbutton");

    await user.click(input);
    await user.keyboard("99");
    await user.click(plusButton);

    expect(input.value).toMatch("99");
  });
});
