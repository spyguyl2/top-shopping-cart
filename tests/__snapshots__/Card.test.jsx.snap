// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > Renders component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dDvxFM PHluN"
      >
        <h4 />
        <img
          alt=""
        />
        <p>
          $undefined
        </p>
        <div
          class="sc-KXCwU fCBWIZ"
        >
          <label>
            Quantity:
          </label>
          <div
            class="sc-gyycJP hJHhVz"
          >
            <button
              aria-label="minus"
              class="sc-eaUbBy bfyhJY"
            >
              <svg
                class="lucide lucide-minus"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
              </svg>
            </button>
            <input
              class="sc-gaZyOd Gexjg"
              max="99"
              min="0"
              type="number"
              value="0"
            />
            <button
              aria-label="plus"
              class="sc-eaUbBy bfyhJY"
            >
              <svg
                class="lucide lucide-plus"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12h14"
                />
                <path
                  d="M12 5v14"
                />
              </svg>
            </button>
          </div>
        </div>
        <button>
          Add To Cart
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dDvxFM PHluN"
    >
      <h4 />
      <img
        alt=""
      />
      <p>
        $undefined
      </p>
      <div
        class="sc-KXCwU fCBWIZ"
      >
        <label>
          Quantity:
        </label>
        <div
          class="sc-gyycJP hJHhVz"
        >
          <button
            aria-label="minus"
            class="sc-eaUbBy bfyhJY"
          >
            <svg
              class="lucide lucide-minus"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
            </svg>
          </button>
          <input
            class="sc-gaZyOd Gexjg"
            max="99"
            min="0"
            type="number"
            value="0"
          />
          <button
            aria-label="plus"
            class="sc-eaUbBy bfyhJY"
          >
            <svg
              class="lucide lucide-plus"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 12h14"
              />
              <path
                d="M12 5v14"
              />
            </svg>
          </button>
        </div>
      </div>
      <button>
        Add To Cart
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
